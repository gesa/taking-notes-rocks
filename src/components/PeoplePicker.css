.countdown-button {
  display: block;
  height: 270px;
  width: 270px;
  margin: 0 auto;
  transition: height 0.2s ease-in-out, width 0.2s ease-in-out;
  visibility: visible;
}

.countdown-button:hover {
  cursor: pointer;
  height: 280px;
  width: 280px;
}

.countdown-button.animate {
  animation: spin 0.6s linear infinite;
}

.countdown-button.fade {
  margin: auto;
  bottom: 0;
  left: 0;
  opacity: 0.5;
  position: absolute;
  right: 0;
  top: 0;
  z-index: -1;
}

@keyframes spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}

