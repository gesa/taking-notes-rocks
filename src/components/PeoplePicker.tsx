import React, { Dispatch, useState } from "react";
import "./PeoplePicker.css";
import { MutateAction, People } from "./types";

export default ({
  dispatch,
  people,
}: {
  dispatch: Dispatch<MutateAction>;
  people: People;
}) => {
  const suspenseInSeconds = 3;
  const [pencilClass, setPencilClass] = useState("");
  const [selectedPerson, setSelectedPerson] = useState("");
  const [resultVisibility, setResultVisibility] = useState("hidden");
  const names = [...people.keys()];

  const doCountdown = () => {
    const eligiblePeople = names.filter((person) => {
      if (people.get(person)) return person;

      return false;
    });

    setPencilClass("animate");

    setTimeout(() => {
      if (eligiblePeople.length === 0) {
        setSelectedPerson("");
      } else {
        const chosenOne =
          eligiblePeople[Math.floor(Math.random() * eligiblePeople.length)];
        setSelectedPerson(chosenOne);
        dispatch({ action: "disable", person: chosenOne });
      }

      setPencilClass("fade");
      setResultVisibility("");
    }, suspenseInSeconds * 1000);
  };

  const Result = () => {
    let resultCopy;

    if (resultVisibility === "hidden") {
      resultCopy = "";
    } else if (selectedPerson === "") {
      resultCopy = (
        <h2>
          Oh no!
          <br />
          No eligible names.
          <br />
          Try again.
        </h2>
      );
    } else {
      resultCopy = (
        <>
          <h2>Grab a pencil</h2>
          <h1>{selectedPerson}</h1>
          <h2>you're taking notes!</h2>
        </>
      );
    }

    return (
      <div className={`countdown-label ${resultVisibility}`}>
        {resultCopy}
        <button
          className="reset-button"
          onClick={() => {
            setSelectedPerson("");
            setResultVisibility("hidden");
            setPencilClass("");
          }}
        >
          Reset
        </button>
      </div>
    );
  };

  return (
    <main
      className="flex-center"
      onClick={() => {
        dispatch({ action: "close-list" });
      }}
    >
      <Result />
      <img
        className={`countdown-button ${pencilClass}`}
        src="/pencil.png"
        alt="An illustration of a pencil"
        onClick={doCountdown}
      />
    </main>
  );
};
