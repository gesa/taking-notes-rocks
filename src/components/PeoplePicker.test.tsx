import React from "react";
import { render, fireEvent, act } from "@testing-library/react";
import PeoplePicker from "./PeoplePicker";

describe("PeoplePicker", () => {
  const testPerson = "John Appleseed";
  const testDisabledPerson = "John Q. Public";
  const testPeople = new Map([
    [testPerson, true],
    [testDisabledPerson, false],
  ]);
  const dispatch = jest.fn();

  beforeEach(() => {
    jest.useFakeTimers();
  });

  it("displays the name if there's only one eligible name", () => {
    const { getByAltText, container } = render(
      <PeoplePicker
        dispatch={dispatch}
        people={new Map([[testPerson, true]])}
      />
    );

    act(() => {
      fireEvent.click(getByAltText(/pencil/));
      jest.runAllTimers();
    });

    expect(container).toHaveTextContent(
      "Grab a pencilJohn Appleseedyou're taking notes!"
    );
  });

  it("displays an eligible name from the list", () => {
    const multiplePeople = new Map([
      [testPerson, true],
      ["John Q. Public", true],
      ["Jane Doe", true],
      ["Red Herring", false],
    ]);
    const { getByAltText, container } = render(
      <PeoplePicker dispatch={dispatch} people={multiplePeople} />
    );

    act(() => {
      fireEvent.click(getByAltText(/pencil/));
      jest.runAllTimers();
    });

    const selectedName =
      container.getElementsByTagName("h1")[0].textContent || "";

    expect(container.getElementsByTagName("div")[0].textContent).toMatch(
      `Grab a pencil${selectedName}you're taking notes!`
    );
    expect(multiplePeople.get(selectedName)).toBeTruthy();
  });

  it("displays error message if there are no eligible names", () => {
    const { getByAltText, container } = render(
      <PeoplePicker
        dispatch={dispatch}
        people={new Map([[testDisabledPerson, false]])}
      />
    );

    act(() => {
      fireEvent.click(getByAltText(/pencil/));
      jest.runAllTimers();
    });
    expect(container).toHaveTextContent("Oh no!No eligible names.Try again.");
  });

  it("displays error message if there are no names at all", () => {
    const { getByAltText, container } = render(
      <PeoplePicker dispatch={dispatch} people={new Map()} />
    );

    act(() => {
      fireEvent.click(getByAltText(/pencil/));
      jest.runAllTimers();
    });
    expect(container).toHaveTextContent("Oh no!No eligible names.Try again.");
  });

  it("uses 3 seconds of suspense", () => {
    const { getByAltText, container } = render(
      <PeoplePicker dispatch={dispatch} people={testPeople} />
    );
    const initialText = container.textContent;

    act(() => {
      fireEvent.click(getByAltText(/pencil/));
      jest.advanceTimersByTime(1000);
      expect(container.textContent).toBe(initialText);
      jest.runAllTimers();
    });

    expect(container.textContent).not.toBe(initialText);
  });

  it("matches snapshot", () => {
    const { getByAltText, container } = render(
      <PeoplePicker dispatch={dispatch} people={testPeople} />
    );

    expect(container).toMatchSnapshot();

    act(() => {
      fireEvent.click(getByAltText(/pencil/));
      jest.runAllTimers();
    });

    expect(container).toMatchSnapshot();
  });
});
