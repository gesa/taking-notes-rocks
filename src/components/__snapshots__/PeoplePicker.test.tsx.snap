// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PeoplePicker matches snapshot 1`] = `
<div>
  <main
    class="flex-center"
  >
    <div
      class="countdown-label hidden"
    >
      
      <button
        class="reset-button"
      >
        Reset
      </button>
    </div>
    <img
      alt="An illustration of a pencil"
      class="countdown-button "
      src="/pencil.png"
    />
  </main>
</div>
`;

exports[`PeoplePicker matches snapshot 2`] = `
<div>
  <main
    class="flex-center"
  >
    <div
      class="countdown-label "
    >
      <h2>
        Grab a pencil
      </h2>
      <h1>
        John Appleseed
      </h1>
      <h2>
        you're taking notes!
      </h2>
      <button
        class="reset-button"
      >
        Reset
      </button>
    </div>
    <img
      alt="An illustration of a pencil"
      class="countdown-button fade"
      src="/pencil.png"
    />
  </main>
</div>
`;
